src/detectors/builtin/suspiciousMessageMode.ts

```diff
@@ -0,0 +1,104 @@
+import { CompilationUnit } from "../../internals/ir";
+import { forEachExpression } from "../../internals/tact";
+import { MistiTactWarning, Severity } from "../../internals/warnings";
+import { ASTDetector } from "../detector";
+import {
+  AstExpression,
+  AstStructInstance,
+  idText,
+  AstOpBinary,
+  AstId,
+} from "@tact-lang/compiler/dist/grammar/ast";
+
+export class SuspiciousMessageMode extends ASTDetector {
+  severity = Severity.MEDIUM;
+
+  async check(cu: CompilationUnit): Promise<MistiTactWarning[]> {
+    const warnings: MistiTactWarning[] = [];
+    Array.from(cu.ast.getProgramEntries()).forEach((node) => {
+      forEachExpression(node, (expr) => {
+        if (this.isSendParametersStruct(expr)) {
+          this.checkSendParameters(expr, warnings);
+        }
+      });
+    });
+    return warnings;
+  }
+  private isSendParametersStruct(expr: AstExpression): boolean {
+    if (expr.kind === "struct_instance") {
+      return idText((expr as AstStructInstance).type) === "SendParameters";
+    }
+    return false;
+  }
+
+  private checkSendParameters(
+    expr: AstExpression,
+    warnings: MistiTactWarning[],
+  ): void {
+    const args = (expr as AstStructInstance).args;
+    const modeField = args.find((arg) => idText(arg.field) === "mode");
+    if (modeField) {
+      this.checkModeExpression(modeField.initializer, warnings);
+    }
+  }
+
+  private checkModeExpression(
+    expr: AstExpression,
+    warnings: MistiTactWarning[],
+  ): void {
+    const flagsUsed = new Set<string>();
+    const traverse = (expr: AstExpression): void => {
```

Comment:
We should use `forEachExpression` here in order to cover corner cases like these:
```
~((x + y) | z)
generateParams(x + y)
```